### DISCUSSION TITLE
ESP32 turning on and off aioble server

### DISCUSSION THREAD

**Discussion by danman009** [micropython/micropython#13491] [ESP32]
Hey all - running into an issue trying to use BLE and WiFi together on my ESP32. From a high level, I want my board to be able to advertise data via a BLE server and also be able to periodically send some data in a WiFi request in a couple of different async loops. I have each of the two loops working the way I want when running them in isolation, but combining them together is where things are getting a bit tricky for me. 

From my understanding, the ESP32 uses the same radio to transmit bluetooth and WiFi, so you actually are not able to connect to either one while the other is active. My issue is with trying to turn off BLE server advertising, connect to WiFi, disconnect from WiFi, and then restart the BLE advertising. 

```
async def start(self, isFirstStart: bool):
        self._advertise = True
        print('GETTING HERE')
        aioble.register_services(self._bluetooth_service)
        # Start up the advertise task in a different async function so that it stays open but we can return
        if isFirstStart:
            print('OVER HERE!')
            self._advertising_task = asyncio.create_task(self.advertise())
        else:
            print('ACTUALLY JUST RUN')
            self._advertising_task.run()
        print('ACUALLY HERE')

def stop_service(self):
        self._advertise = False
        if self._advertising_task:
            self._advertising_task.cancel()
            self._advertising_task = None
            aioble.stop()
        for characteristic in self._characteristics:
            characteristic.stop()
        for task in self._tasks:
            task.cancel()
        self._tasks = []
        print('DONE STOPPING')

        
async def advertise(self):
    while self._advertise:
        try:
            connection = await aioble.advertise(
                interval_us=250000,
                name="Bloom",
                services=[self._DEFAULT_ENVIRONMENT_SENSOR_UUID],
                appearance=self._DEFAULT_ENVIRONMENT_SENSOR_APPEARANCE_CONST,
                manufacturer=(0xabcd, b"1234")
            )
            print("Connection from", connection.device)
        except Exception as e:
            print("Error during advertisement:", e)
            await asyncio.sleep(1)  # Optional: wait a bit before retrying
        finally:
            if not self._advertise:
                aioble.stop()
                print('AIOBLE STOP')
            print('EXITING FINALLY')
```

My code snippets are above. From my testing, it seems like I am able to successfully initially start the server, turn it off via aioble.stop(), connect to wifi, disconnect from wifi, but when trying to restart the ble server in advertise(), I am getting unexpected behavior, and the error is this:

Error during advertisement: [Errno 116] ETIMEDOUT
EXITING FINALLY
E (25130) BLE_INIT: controller init failed

Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

I'm fairly new to this and just looking to sanity check that I am on the right path. Is aioble.stop() the correct way to free up the bluetooth/wifi radio for my goal? Or am I going completely off in the wrong direction.

**Comment by bixb922** [comment]
The ESP32 works well with both BLE and WiFi active and transmitting/receiving. No need to turn one off to use the other.

**Reply by nebelgrau77** [reply]
This is unfortunately not true. In my case on ESP32C3 with MicroPython 1.24 an active WiFi connection prevents BLE connections. But if I disconnect from WiFi, then BLE works without a problem. 
I found this thread because of similar issues that the OP is facing. In my case I'm trying to figure out how to correctly handle failed BLE connections, e.g. when the peripheral device does not answer (my Micropython ESP32 device acting as central).

**Reply by bixb922** [reply]
There is a ESP32 family that comprises the several microcontrollers: ESP32 (the original model),  ESP32-S3, ESP32-C3, ESP32-S2, ESP32-C6 and others. Although they are sold under the same ESP32 brand name, the individual microcontroller architectures and characteristics sometimes differ substantially.

I have seen WIFI and Bluetooth coexist well at least  on the original ESP32 and the ESP32-S3.

The ESP32-C3 has hardware restrictions that other models don't necessarily have, see https://docs.espressif.com/projects/esp-idf/en/v5.0/esp32c3/api-guides/coexist.html

This is a hardware characteristic, not a MicroPython issue. 

I believe the OP refers to the ESP32 microcontroller, not the complete family.

I hope that helps.
